### ISSUE TITLE
statnet suite

### ISSUE THREAD

**Issue by knapply** [thomasp85/tidygraph#64]
Thank you for the work on this (as well as ggraph, and patchwork!). Seriously awesome stuff that I find myself using (and recommending) more and more.

I'm developing material based on lessons-learned after transitioning a network analysis course to R from UCINET, Pajek, etc. This includes abstracting away the necessary jumps between frameworks so students spend less time confused by code and more time learning the material.

Is the plan to rely solely on igraph and igraph-based extenstions for calculations, or are is there a plan to integrate capabilities, metrics, models, etc. from the statnet suite and elsewhere? 

tl;dr: Where do you plan on tidygraph heading in the future?

**Comment by thomasp85** [comment]
That is a pretty big question🙂 I’ll try to answer below but keep in mind that any roadmap is always under the rule of my time and priorities:

- the biggest missing piece in `tidygraph` IMO is the modelling part. This is partly because my own use and interest has laid closer to algorithms and data structures. I’m not going to provide an alternative StatNET backend but I do plan to somehow interface with their packages such as `ergm`. My biggest problem is lack of experience in this field which makes my less certain about designing the API
- other `tidygraph` related plans are to provide specialised classes and methods for spatial and temporal network
- I have continuously toyed with the idea of providing `dbplyr` like functionality that allows the user to query graph databases such as GraphX and Neo4J using `tidygraph` API
- I really want to build an interface to SNAP and then provide it as a highly performant alternative backend

Hope this answers your questions

**Comment by syknapptic** [comment]
Thank you! Sorry for the size of the question. I should have prefaced it by saying that I'm working on things that I'd like to contribute to `tidygraph`, but wanted to see if they were within scope or even of interest. I came to R for networks, `ergm`s specifically, but I stayed for the `tidyverse`. Your work here really resonates with me.

It's unfortunate that `igraph` and `network` objects aren't more compatible and I continue to run across cases where `intergraph::as*` or `as_tbl_graph()` fail because of some odd aspect of a `network` object that requires manual tinkering.

That said, there are many functions on the `statnet` side that don't seem to exist elsewhere (brokerage roles, quadratic assignment problems, ergm, tergm, etc.). I'm curious if either of these sound reasonable as contributions:

* Improve conversion methods to be able to swap back and forth and wrap desired `statnet` functions in a way that returns results in a format consistent with `tidygraph`
* Improve conversion methods, but keep functions wrapping `statnet` in a separate package as a `tidygraph` extension and prevent an additional pile of dependencies.

Since `statnet` is a monster spanning ~12 packages (swaths of which are written in C), those _seem_ to be the most reasonable options, but both have issues. I wouldn't claim to be a developer expert here (entirely internal-use stuff), so I would really appreciate your thoughts.

As for your other plans, I __really__ like the ideas for spatial and temporal classes and something akin to `dbplyr`. As for a high-performance backend, I remember reading a paper referring to work being done to write a network engine in Edward (tensorflow-based) to capitalize on GPU capabilities and greatly reduce the pain of `ergm`s. If you're curious, I'll look for the paper and pass it along.

Thanks, and cheers!

**Comment by thomasp85** [comment]
I would welcome any contributions, but for larger stuff I think a discussion in separate issues will ensure that we are on the same page. As a starter, if you have interest in improving the conversion to and from network I would be more than happy. As to where a statnet interface should reside, in unsure at the moment. Tidygraph have a lot of stuff in Suggest to avoid dependency hell for users that don’t need the functionality so this could work for statnet as well. It all depends on the size of the challenge. Let’s keep that question open for now.

**Comment by knapply** [comment]
I now question the practicality here. The conversion solutions can be rather hacky depending on the network. So far, it’s always cleaner to build whatever’s missing from igraph from scratch and run it on the raw matrices.

With that in mind, I think a purpose-built stats backend seems like a better time investment.

**Comment by thomasp85** [comment]
Do you have any feeling with how much work it would require to reimplement the most important models from statnet?

I would generally like to avoid it as that would sever ties with a very healthy community...

**Comment by knapply** [comment]
I don’t think that came across as intended.

All parties would appreciate a meta class that can standardize representations, maximize extensibility, and is just plain faster than whatever they’re using. 

Some models (QAP, CUG come to mind) are not terrible to build, but ERGM and friends are different animals. They’re very expensive computationally expensive with anything but small networks and simple model terms. I’ve never seen a framework other than statnet that implements them. graph-tool, what seems to be the hotness in Python-land these days, doesn’t even seem to have them.

I’m not at all suggesting that ties be severed; I’m more insisting they be grown.

I wish this were more immediately helpful, but I’ll contribute some conversion routines for review once I’m more confident in they’re handling of bipartite networks.
